part/node.go

```diff
@@ -258,47 +263,154 @@ func (n *header[T]) children() []*header[T] {
 }
 
 func (n *header[T]) findIndex(key byte) (*header[T], int) {
-	if n.kind() == nodeKind256 {
+	switch n.kind() {
+	case nodeKindLeaf:
+		return nil, 0
+	case nodeKind4:
+		n4 := n.node4()
+		size := n4.size()
+		for i := 0; i < int(size); i++ {
+			if n4.keys[i] == key {
+				return n4.children[i], i
+			} else if n4.keys[i] > key {
+				return nil, i
+			}
+		}
+		return nil, size
+	case nodeKind16:
+		n16 := n.node16()
+		size := n16.size()
+		for i := 0; i < int(size); i++ {
+			if n16.keys[i] == key {
+				return n16.children[i], i
+			} else if n16.keys[i] > key {
+				return nil, i
+			}
+		}
+		return nil, size
+	case nodeKind48:
+		children := n.children()
+		idx := sort.Search(len(children), func(i int) bool {
+			return children[i].prefix[0] >= key
+		})
+		if idx >= n.size() || children[idx].prefix[0] != key {
+			// No node found, return nil and the index into
+			// which it should go.
+			return nil, idx
+		}
+		return children[idx], idx
+	case nodeKind256:
 		return n.node256().children[key], int(key)
+	default:
+		panic(fmt.Sprintf("unknown node kind: %x", n.kind()))
 	}
-
-	children := n.children()
-	idx := sort.Search(len(children), func(i int) bool {
-		return children[i].prefix[0] >= key
-	})
-	if idx >= n.size() || children[idx].prefix[0] != key {
-		// No node found, return nil and the index into
-		// which it should go.
-		return nil, idx
-	}
-	return children[idx], idx
 }
 
 func (n *header[T]) find(key byte) *header[T] {
-	child, _ := n.findIndex(key)
-	return child
+	switch n.kind() {
+	case nodeKindLeaf:
+		return nil
+	case nodeKind4:
+		n4 := n.node4()
+		size := n4.size()
+		for i := 0; i < int(size); i++ {
+			if n4.keys[i] == key {
+				return n4.children[i]
+			} else if n4.keys[i] > key {
+				return nil
+			}
+		}
+		return nil
+	case nodeKind16:
+		n16 := n.node16()
+		size := n16.size()
+		for i := 0; i < int(size); i++ {
+			if n16.keys[i] == key {
+				return n16.children[i]
+			} else if n16.keys[i] > key {
+				return nil
+			}
+		}
+		return nil
+	case nodeKind48:
+		n48 := n.node48()
+		idx := n48.index[key]
+		if idx < 0 {
+			return nil
+		}
+		return n48.children[idx]
+	case nodeKind256:
+		return n.node256().children[key]
+	default:
+		panic("unknown node kind")
```

Comment:
nit: print unknown kind like in `findIndex`

Reply:
went through all switches and unified the default cases.